### DISCUSSION TITLE
Is it possible to have a .part file locally repaired to mp4

### DISCUSSION THREAD

**Discussion by creativestate** [aclap-dev/video-downloadhelper#666] [Help and Q&A] [answered]
### Your question

I have a downloaded file which appears to have been downloaded successfully as the rest of the video files have completed successfully (a while back), one of the files is in a .part format, which clearly indicates that it did not complete the conversion process. The file appears "complete" state in terms of file size, given the file is now downloaded am I able to repair/complete this process to get it in desired readable mp4 format? 

The online source of this file is no longer available, so I cannot just reattempt the download. The file was downloaded sometime last year, so it was using the previous coapp version of 1.6 or something along those lines and not the updated version shown below.

### Addon & CoApp version and details

version: 9.0.1.22
target: google
channel: stable
lang: en-US
coapp: {"found":true,"new_version":false,"path":"C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe","version":"2.0.19"}
license: {"unset":true}
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36
{
    "http_media_download_strategy": "inbrowser",
    "used_sidebar_button": true,
    "used_popup_button": true,
    "download_directory": "",
    "view_options": {
        "all_tabs": true,
        "hide_downloaded": false,
        "low_quality": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": true,
        "sort_by_status": true,
        "sort_reverse": false
    },
    "use_wide_ui": true,
    "use_legacy_ui": false,
    "use_sidebar": true,
    "open_count_store": 1
}

**Comment by paulrouget** [comment]
My only idea is to use VLC to do so.

**Reply by creativestate** [reply]
Right, would you be able to tell me once a download is finished, its expected to be in the .part format? If so, is it just a container file for download, like a separated version of the audio and video file that needs remuxing?

Why would I use VLC, what am I attempting to achieve there? VLC cannot open the file for playback if that is what you are suggesting; I tried removing the .part extension but it refuses to play back, so I am assuming something still needs to happen to the file after the download with your application/extension which did not happen when it downloaded the file originally, but the rest did.

**Reply by GrampaWildWilly** [reply]
VLC?  I wouldn't recommend that.  I would recommend trying to use VDH's own Convert local files function.  But in general, if you are left with a .part file, that does not mean simply that the download succeeded but conversion failed.  It means the download itself failed.  You say VLC can't play it.  Does it generate any errors in its log?  My expectation is it would say, "moov atom not found."  I'm guessing you were trying to download an MP4, or at least convert to MP4.  The moov atom is uniquely an MP4 artifact.  If it's missing, it is likely that your .part file is irretrievable.  You might try giving the file an extension of .raw & trying to run it through ffmpeg to try to convert it to something else.  But ffmpeg usually gags on files that don't have their moov atoms.  I fear you are simply out of luck.  Too bad you can't just download it again.

**Reply by paulrouget** [reply]
> Right, would you be able to tell me once a download is finished, its expected to be in the .part format?

We download into a normal file. If the download ends properly, or if you click "stop", then the file is supposed to be valid. If the download gets interrupted for other reasons, then the file needs "reindexing", which VLC will do if you open that specific file.

There no .part files.

> If so, is it just a container file for download, like a separated version of the audio and video file that needs remuxing?

No. Remuxing is happening in memory, no on the disk. We "mux before write".

> Why would I use VLC, what am I attempting to achieve there? VLC cannot open the file for playback if that is what you are suggesting; I tried removing the .part extension but it refuses to play back, so I am assuming something still needs to happen to the file after the download with your application/extension which did not happen when it downloaded the file originally, but the rest did.

If you have a .part file, that's from a much older version of the extension, and to be honest, I am not exactly sure what's in that part file (depends on how the download was triggered). ffprobe will tell you what's in there, and ffmpeg or vlc will help you either re-index it or remux it.

**Reply by GrampaWildWilly** [reply]
He did say his download was from rather a long time ago on an older version of VDH.  Used to be VDH would create a file with the .part extension & write the downloaded data into it.  When the download completed successfully, the file would be renamed to its proper name.  If I'm not mistaken, this is sort what Firefox downloads do.  But I'm not about to tell you, of all people, how Firefox works.  Also, and I think this was unique to YouTube, there would be 2 .part files in some temporary directory somewhere.  When the downloads of the video & the audio tracks were both completed successfully, VDH would use ffmpeg to merge the 2 tracks & put the result where you wanted, then delete the 2 .part files.  Back in those days, VDH didn't use ffmepg for downloading, only for functions like this merge.

When you say "reindex," is that what ffmpeg would do if you input a .raw file & specified a target file of, let's say, .mp4?  I have to say that I have occasionally tried to do something like this & as often as not, I just got the dreaded "missing moov atom" error.  Sometimes.  Sometimes I've managed to make it work, sometimes not.  I think it very much depends on how damaged the input is.

**Reply by GrampaWildWilly** [reply]
Another thing.  The Stop function didn't used to work correctly.  You fixed it, remember?

**Reply by creativestate** [reply, marked as answer]
Gentlemen, thanks for your feedback. I was able to successfully repair the file from a reference file that had the same encoding settings from the batches of webinar videos that were downloaded at the time.

I was getting the dreaded "moov atom not found /e/broken.mp4.part: Invalid data found when processing input", when using ffmpeg on the file.

I used untrunc from https://github.com/ponchio/untrunc, with command: "untrunc /e/reference.mp4 /e/broken.mp4.part" and this gave me a fully playable webinar file again.

Appreciate all the information and your time.
